### PR TITLE
Issue 36

### PR DESCRIPTION
**Pull request by awasthiy**
Enhancement #36 

Introducing new feature request : Running Test File from CLI 

Introduced a new cli option "--testfiles" where along with these options "--hostname","--login" and "--passwd"  instead of config file ( no need of config file ). You can provide multiple test files

Example:

For single test file
```
jsnapy --snapcheck POST --testfiles test_is_gt.yml --hostname 10.221.130.68 --login abc -passwd xyz
```
or

```
jsnapy --snapcheck POST -T test_is_gt.yml -t 10.221.130.68 -l abc -p xyz
```

Following output is produced:

```
Connecting to device 10.221.130.68 ................
Taking snapshot of COMMAND: show chassis fpc 
*************************** Device: 10.221.130.68 ***************************
Tests Included: check_chassis_fpc 
************************* Command: show chassis fpc *************************
PASS | All "cpu-total" is greater than 2" [ 1 matched ]
------------------------------- Final Result!! -------------------------------
check_chassis_fpc : Passed
Total No of tests passed: 1
Total No of tests failed: 0 
Overall Tests passed!!! 
```
For Multiple test files
```
jsnapy --snapcheck POST --testfiles test_exists.yml test_is_get.yml test_exists2.yml --hostname 10.221.130.68 --login abc -passwd xyz
```
Following Output is Produced
```
Connecting to device 10.221.142.171 ................
Taking snapshot of COMMAND: show version 
Taking snapshot of COMMAND: show chassis fpc 
Taking snapshot of COMMAND: show chassis alarms 
*************************** Device: 10.221.142.171 ***************************
Tests Included: test_version_check 
*************************** Command: show version ***************************
PASS | All "//package-information/name" exists at xpath "//software-information" [ 52 matched ]
*************************** Device: 10.221.142.171 ***************************
Tests Included: check_chassis_fpc 
************************* Command: show chassis fpc *************************
ERROR!! Node <cpu-total> not found at xpath <//fpc[normalize-space(slot) = "0"]> for IDs: {}
FAIL | All "cpu-total" is not greater than  "1" [ 0 matched / 1 failed ]
*************************** Device: 10.221.142.171 ***************************
Tests Included: check_chassis_alarms 
************************ Command: show chassis alarms ************************
PASS | All "no-active-alarms" exists at xpath "//alarm-information/alarm-summary" [ 1 matched ]
------------------------------- Final Result!! -------------------------------
check_chassis_alarms : Passed
check_chassis_fpc : Failed
test_version_check : Passed
Total No of tests passed: 2
Total No of tests failed: 1 
Overall Tests failed!!!
```

### CODE REVIEW

File: lib/jnpr/jsnapy/jsnapy.py [hunk 1]
```diff
@@ -312,19 +317,32 @@ def get_hosts(self):
...
-                "ERROR!! Config file '%s' is not present " %
-                conf_file, extra=self.log_detail)
-            sys.exit(1)
+            if self.args.hostname and self.args.login and self.args.passwd:
+                temp_dict = {'hosts':[{'device':'', 'username':'', 'passwd':''}], 'tests':[]}
+                temp_dict['hosts'][0]['device'] = self.args.hostname
+                temp_dict['hosts'][0]['username'] = self.args.login
+                temp_dict['hosts'][0]['passwd'] = self.args.passwd
+                temp_dict['tests'].append(self.args.test_file)
+                with open('result.yml', 'w') as yaml_file:
```
Comment (eeishaan) [inline lib/jnpr/jsnapy/jsnapy.py]: Writing to a file is not required. yaml.load() returns a dict and as long as your variable temp_dict is in line with expected dict format, you can assign it to self.main_file

File: tests/unit/test_jsnapy.py [hunk 2]
```diff
@@ -728,10 +728,141 @@ def test_port_with_include(self,mock_connect,mock_path):
...
+        js.args.diff = False
+        js.args.login = "regress"
+        js.args.hostname = "10.221.130.68"
+        js.args.passwd = "MaRtInI"
+        js.args.post_snapfile = None
+        js.args.pre_snapfile = "mock_snap"
+        with patch('argparse.ArgumentParser.print_help') as mock_parser:
+            js.check_arguments()
+            js.get_hosts()
+            self.assertTrue(js.args.file)
```
Comment (eeishaan) [inline tests/unit/test_jsnapy.py]: Same improvements as above

File: tests/unit/test_jsnapy.py [hunk 3]
```diff
@@ -728,10 +728,141 @@ def test_port_with_include(self,mock_connect,mock_path):
...
+        js.args.diff = False
+        js.args.login = "regress"
+        js.args.hostname = "10.221.130.68"
+        js.args.passwd = "MaRtInI"
+        js.args.post_snapfile = "mock_snap2"
+        js.args.pre_snapfile = "mock_snap"
+        with patch('argparse.ArgumentParser.print_help') as mock_parser:
+            js.check_arguments()
+            js.get_hosts()
+            self.assertTrue(js.args.file)
```
Comment (eeishaan) [inline tests/unit/test_jsnapy.py]: Make this check more strict and check what is the expected value of args.file . Do a string check

File: tests/unit/test_jsnapy.py [hunk 4]
```diff
@@ -728,10 +728,141 @@ def test_port_with_include(self,mock_connect,mock_path):
...
```
Comment (eeishaan) [inline tests/unit/test_jsnapy.py]: Same improvement here as well

File: tests/unit/test_jsnapy.py [hunk 5]
```diff
@@ -728,10 +728,141 @@ def test_port_with_include(self,mock_connect,mock_path):
...
+            self.assertTrue(js.args.file)        
+
+
+    @patch('argparse.ArgumentParser.exit')
+    @patch('jnpr.jsnapy.SnapAdmin.get_hosts')
+    @patch('jnpr.jsnapy.jsnapy.sys.exit')
+    @patch('jnpr.jsnapy.jsnapy.logging.getLogger')
+    def test_check_arguments_test_file_4(self, mock_log,mock_sys, mock_get_hosts, mock_arg):
+        argparse.ArgumentParser.parse_args = MagicMock()
+        argparse.ArgumentParser.parse_args.return_value = argparse.Namespace(check=False,
```
Comment (eeishaan) [inline tests/unit/test_jsnapy.py]: I think you can remove line 818, 819. Investigate this and remove if unnecessary

File: lib/jnpr/jsnapy/jsnapy.py [hunk 6]
```diff
@@ -312,19 +317,28 @@ def get_hosts(self):
...
+                fpath = get_path('DEFAULT', 'config_file_path')
+                config_file = open(os.path.join(fpath, conf_file), 'r')
+                self.main_file = yaml.load(config_file)
         else:
-            self.logger.error(
-                colorama.Fore.RED +
-                "ERROR!! Config file '%s' is not present " %
-                conf_file, extra=self.log_detail)
-            sys.exit(1)
+            if self.args.hostname and self.args.login and self.args.passwd:
```
Comment (vnitinv) [inline lib/jnpr/jsnapy/jsnapy.py]: shouldn't we check for self.args.test_file also?

File: lib/jnpr/jsnapy/jsnapy.py [hunk 7]
```diff
@@ -1049,9 +1063,9 @@ def check_arguments(self):
             self.parser.print_help()
             sys.exit(1)
 
-        if((self.args.snap is True and (self.args.pre_snapfile is None or self.args.file is None)) or
+        if(((self.args.snap is True and (self.args.pre_snapfile is None or self.args.file is None)) or
             (self.args.snapcheck is True and self.args.file is None) or
-            (self.args.check is True and self.args.file is None)
+            (self.args.check is True and self.args.file is None)) and (self.args.test_file is None or self.args.hostname is None or self.args.login is None or self.args.passwd is None)
```
Comment (vnitinv) [inline lib/jnpr/jsnapy/jsnapy.py]: You code violates pep8 (line not more than 80 char, can we indent code properly.

File: lib/jnpr/jsnapy/jsnapy.py [hunk 8]
```diff
@@ -1067,7 +1081,7 @@ def check_arguments(self):
                     None)
                 sys.exit(1)
             else:
-                if self.args.file is None:
+                if self.args.file is None and (self.args.test_file is None or self.args.hostname is None or self.args.login is None or self.args.passwd is None):
```
Comment (vnitinv) [inline lib/jnpr/jsnapy/jsnapy.py]: You code violates pep8 (line not more than 80 char, can we indent code properly.

File: lib/jnpr/jsnapy/jsnapy.py (hunk 6 above)
Comment (vnitinv) [inline lib/jnpr/jsnapy/jsnapy.py]: self.args.login and self.args.passwd are optional, so you should check them.
Say if user has configured ssh, they dont need to pass password. Please remove that

File: lib/jnpr/jsnapy/jsnapy.py [hunk 9]
```diff
@@ -110,6 +110,11 @@ def __init__(self):
             help="post snapshot filename",
             type=str)       # make it optional
         self.parser.add_argument(
+            "-test", "--test_file",
```
Comment (vnitinv) [inline lib/jnpr/jsnapy/jsnapy.py]: change -test to -T as discussed.

--test_file to --testfiles

File: lib/jnpr/jsnapy/jsnapy.py [hunk 10]
```diff
@@ -110,6 +110,11 @@ def __init__(self):
...
+            nargs='?',
+            help="test filename",
```
Comment (vnitinv) [inline lib/jnpr/jsnapy/jsnapy.py]: Give proper discription and add docstring for this parameter.

**Comment by awasthiy** [comment]
All requested changes are done

**Pull request merged**